Improved Chrome audit performance

ryanachi opened the pull request:
I've added service worker and web app manifest files, along with editing the Makefile and index.html.in to reflect these changes.  The 'Progressive Web App' rating is now at 73, while the 'Performance' is at 71.

The following errors have been resolved:
-Does not register a service worker
-Uses HTTP instead of HTTPS
-Configured for a custom splash screen
-Address bar matches brand colors
-Page load is fast enough on 3G

The following errors still remain: 
![revised chrome audits](https://user-images.githubusercontent.com/34379340/33779498-03b89870-dc02-11e7-8c4d-bb4f9484af3d.png)

sushain97 commented on Makefile:
```diff
@@ -236,4 +240,4 @@ server:
 ### Clean ###
 clean:
-	rm -rf build/
+	rm -rf build/
```
Why was this newline removed?

sushain97 commented on Makefile:
```diff
@@ -43,7 +43,8 @@ JSFILES= \
 	assets/js/generator.js \
-	assets/js/sandbox.js
+	assets/js/sandbox.js \
+    assets/js/sw.js
```
Indentation?

sushain97 replied:
Can this be removed now?

sushain97 commented:
@ryanachi the build is still failing and there are a number of HoundCI errors visible above.

ryanachi commented:
The following issues have been resolved:

-Uses HTTP instead of HTTPS (all of the HTTP links have been changed to HTTPS)
-Configured for a custom splash screen (a web app manifest has been added, along with app icons in multiple sizes)
-Address bar matches brand colors (both the web app and main page now have matching colors)
-Page load is fast enough on 3G (I haven’t changed anything to fix this issue, but it doesn’t apply anymore)

The following issues still remain:

-does not register a service worker (not fixed because the service worker would be largely unnecessary)
-does not respond with a 200 when offline (requires a registered service worker to pass, same reason as above)
-user will not be prompted to install a web app (requires a registered service worker to pass, same reason as above)
-does not redirect HTTP traffic to HTTPS (requires adding canonical links to the <head> section of all HTML pages of the website and configuring the server to redirect all HTTP links received; perhaps this should be dealt with separately)

sushain97 commented:
> The following issues still remain:

What about performance, accessibility and best practices?

sushain97 commented on index.html.in:
```diff
@@ -2,6 +2,7 @@
 
 <html>
     <head>
+        <meta name="theme-color" content="#2196F3"/>
```
How did you pick this color?

ryanachi replied:
I selected this color as the Google Developers recommended color; although, I'd be happy to switch it to any other one.  Choosing a color is necessary to pass the "Address bar matches brand colors" test.

sushain97 replied:
Ah, I think this is fine.

ryanachi commented:
Due to the removal of the service worker, the audit ratings are now as follows:

Progressive Web App: 64
Peformance: 72
Accessibility: 89
Best Practices: 75

Issues that remain:

**Performance:** 
Reduce render-blocking scripts and stylesheets (need to add defer or async attributes to scripts, add disabled or media attributes to stylesheets, add async attribute to import tags)
Enable text compression (strip out comments, tabs, and spaces)
Serve images as WebP (encode images in WebP instead of JPEG or PNG) 
Keep server response times low 

**Accessibility:**
Make buttons accessible
Add labels for form elements
Ensure that all text has a sufficient contrast ratio
Specify a [lang] attribute for HTML pages

**Best Practices:**

Use HTTP/2 instead of HTTP/1
Add the tag rel="noopener" to links with the tag target="_blank"
Use natural aspect ratio for images

Of the issues, the Accessibility and Best Practices issues seem the least helpful, although they can be fixed.

sushain97 commented:
> Of the issues, the Accessibility and Best Practices issues seem the least helpful, although they can be fixed.

Why do you think these would be the least helpful? I think they would be quite small and good improvements that would be maintainable as well. That's not to say we should do all of them, e.g. "Use natural aspect ratio for images" seems potentially difficult and not too helpful. However, "Add labels for form elements" seems easy and useful for screen readers.

ryanachi commented:
I meant that some of them seemed a bit unnecessary, like the one you pointed out or the text contrast ratio.  However, I agree, some of them would be beneficial adjustments.  Which features should we also implement?

sushain97 commented:
I think it's possible get these fixed:

Performance:
Reduce render-blocking scripts and stylesheets (need to add defer or async attributes to scripts, add disabled or media attributes to stylesheets, add async attribute to import tags)

Accessibility:
Make buttons accessible
Add labels for form elements
Specify a [lang] attribute for HTML pages

Best Practices:
Add the tag rel="noopener" to links with the tag target="_blank"

If one of them turns out to be a lot more work than expected, just say so.

sushain97 commented on Makefile:
```diff
@@ -43,7 +43,7 @@ JSFILES= \
 	assets/js/analyzer.js \
 	assets/js/generator.js \
-	assets/js/sandbox.js
+	assets/js/sandbox.js \
```
I don't think the trailing \ belongs here.

ryanachi commented:
I've updated the code to pass the additional audit tests that we decided on (except 'Reduce render-blocking scripts and stylesheets' as adding a "defer" or "async" tag to stylesheets causes the Bootstrap to fail).  

For the "Best Practices" test, I've added a "rel='noopener'" tag to all of the URLs in index.html.in with the tag "target='_blank'"; however, there are still several URLs outside the index.html.in file that fail the test that I couldn't locate (e.g., there's a link to https://en.wikipedia.org/wiki/Pivot_language).

sushain97 commented on tools/localise-html.py:
```diff
@@ -47,7 +47,7 @@ def handle_starttag(self, tag, attrs):
             if tag == "html":
                 text = self.get_starttag_text()
-                self.p('<html dir="%s"' % ('rtl' if self.localename in rtl_languages else 'ltr') + text[text.index('html') + 4:])
+                self.p('<html dir="%s" lang=en' % ('rtl' if self.localename in rtl_languages else 'ltr') + text[text.index('html') + 4:])
```
I'm pretty sure this should be something like `self.localename`, not always `en`.

sushain97 commented on index.html.in:
```diff
@@ -183,11 +185,11 @@
                                   </div>
                                   <div class="col-sm-6">
-                                      <textarea class="form-control" id="translatedText" rows="15" spellcheck="false" readonly></textarea>
+                                      <textarea class="form-control" id="translatedText" rows="15" spellcheck="false" aria-label="translatedText" readonly></textarea>
```
If I understand correctly, `aria-label` should be proper human readable words. e.g. "Translated Text"

sushain97 replied:
(this comment applies to the other attributes above)

sushain97 commented:
> adding a "defer" or "async" tag to stylesheets causes parts of the CSS to fail

Interesting. Why?

> however, there are still several URLs outside the index.html.in file that fail the test that I couldn't locate

Check `assets/strings/[...].json` files.

ryanachi commented:
I've fixed the aria-labels, changed the [lang] attribute to reflect the self.localename, and added the tag "rel='noopener'" to all of the URLs with the tag "target='_blank'".  (Our new audit scores: Accessibility–97, Best Practices–81. :D)

> Interesting. Why?

In order to pass the audit, either async or defer have to be applied to all of the stylesheets.  When applied to the bootstrap-rtl.min.css, it delays the stylesheet or forces it to load asynchronously to the rest of the page.

sushain97 commented on tools/localise-html.py:
```diff
@@ -47,7 +47,7 @@ def handle_starttag(self, tag, attrs):
             if tag == "html":
                 text = self.get_starttag_text()
-                self.p('<html dir="%s" lang=en' % ('rtl' if self.localename in rtl_languages else 'ltr') + text[text.index('html') + 4:])
+                self.p('<html dir="%s" lang=' % ('rtl' if self.localename in rtl_languages else 'ltr') + self.alpha2name + text[text.index('html') + 4:])
```
It seems like this would be far better as `'<html dir="%s" lang="%s"'`? Your current version is also missing the quotes (it's valid but not great HTML).

sushain97 commented on tools/localise-html.py:
```diff
@@ -105,6 +105,8 @@ def run(html_path, json_path, out_path, conf_path, fallback_path):
     parser.locale = json.loads("".join(open(json_path).readlines()))
     parser.localename = path.basename(json_path).replace('.json', '')
+    parser.alpha2name = parser.locale["alpha2"]
+    print(parser.alpha2name)
```
Extra left over print?

sushain97 commented on index.html.in:
```diff
@@ -368,7 +368,7 @@
               </div>
               <div class="visible-xs">
-                  <select class="localeSelect pull-left" aria-label="pull"><option> </option></select>
+                  <select class="localeSelect pull-left" aria-label="Select Page Language"><option> </option></select>
```
`Select Language` instead should be good

sushain97 commented on index.html.in:
```diff
@@ -54,6 +55,7 @@
                 button#translateDoc { display: none; }
             </style>
         <![endif]-->
+        <link rel="manifest" href="/manifest.json">
```
This should be `./manifest.json`

ryanachi commented:
The errors have been fixed!

> It seems like this would be far better as '<html dir="%s" lang="%s"'? Your current version is also missing the quotes (it's valid but not great HTML).
I actually tried that when I first tried to add the localename; however, the audit only accepts 2-letter ISO codes as valid BCP codes.

> Also, are you certain that manifest.json is getting copied into build? On Circle it looks like it isn't: https://circleci.com/gh/goavki/apertium-html-tools/113#artifacts/containers/0. 
I tested it again, and it looks like the manifest.json file isn't being copied into build.  I'm not sure why this is happening, because I have the following line in Makefile:

### MANIFEST ###
build/manifest.json: assets/manifest.json build/.d
	cp $< $@

Perhaps I've made a syntax error somewhere?

sushain97 commented:
> I actually tried that when I first tried to add the localename; however, the audit only accepts 2-letter ISO codes as valid BCP codes.

Indeed! I meant something like:

    self.p('<html dir="%s" lang="%s"' % ('rtl' if self.localename in rtl_languages else 'ltr', self.alpha2name) + '" '+ text[text.index('html') + 4:])


> I'm not sure why this is happening, because I have the following line in Makefile:

Ah, this isn't sufficient because there's no goal that asks for the file to actually be made. You should add it to `prod`:

    prod: js css html fonts build/sitemap.xml build/manifest.json build/strings/locales.json localhtml images

ryanachi commented:
Ah, that makes much more sense.  I've fixed the (hopefully) final errors, and let me know if more appear.  Thanks!

sushain97 commented:
I ended up making one fix for you and merging it in: https://github.com/goavki/apertium-html-tools/pull/224.

Nice work! Be a little bit more careful with the Python next time :)

ryanachi commented:
Thanks for all of your help!  And yeah, I will. :)

sushain97 commented:
@ryanachi no problem! Feel free to take up other html-tools tasks.

ryanachi commented:
@sushain97 Sounds good!
